### ISSUE TITLE
mutation operators do not scale well

### ISSUE THREAD

**Issue by CoolRunning** [darioizzo/d-CGP#29]
https://github.com/darioizzo/d-CGP/blob/bbef535fa3ddd81e2fcf90faf8215874e5fd0c31/include/dcgp/expression.hpp#L642

In cases where there are a lot of input nodes (but a relatively small network) the "try and discard"-while loops in this area are somewhat inefficient.

**Comment by darioizzo** [comment]
```c++
if (m_active_genes.size() > m_m) {
            for (auto i = 0u; i < N; ++i) {
                unsigned idx = 0u;
                while (idx < m_n) { // we get a random active node (there will be one that is not an input node)
                    idx = m_active_nodes[std::uniform_int_distribution<unsigned>(
                        0, static_cast<unsigned>(m_active_nodes.size() - 1u))(m_e)];
                }
                idx = m_gene_idx[idx] + std::uniform_int_distribution<unsigned>(1, _get_arity(idx))(m_e);
                mutate(idx);
            }
        }
```

I think these lines can be safely changed to something like:
```c++
if (m_active_genes.size() > m_m) {
            for (auto i = 0u; i < N; ++i) {
               idx = m_active_nodes[std::uniform_int_distribution<unsigned>(
                        m_n, static_cast<unsigned>(m_active_nodes.size() - 1u))(m_e)];
                idx = m_gene_idx[idx] + std::uniform_int_distribution<unsigned>(1, _get_arity(idx))(m_e);
                mutate(idx);
            }
        }
```
to be tested.

**Comment by CoolRunning** [comment]
I think the condition `while (idx < m_n)` is always false (apart from the first iteration). `idx` is an element of `m_active_nodes` which only contains indices of active nodes. It seems that input nodes do not count as active as you can easily verify by calling something like `dcgpann.get_active_nodes()` in python. As such the whole while loop has no impact. There must be another reason why I experience these scaling issues.

**Comment by darioizzo** [comment]
I think they do.
```
dcgp = dcgpy.expression_double(1,1,1,10,11,2,dcgpy.kernel_set_double(["sum","diff","mul","div"])(), 32)
dcgp.get_active_nodes()
```
results in 
```
[0,1]
```
on my machine

**Comment by darioizzo** [comment]
So this explains why the rejection method scales badly.
An opion to solve the issue, if needed, is to find the index in m_active_nodes where m_active_nodes[idx]>=m_n and sample starting from there only once

**Comment by CoolRunning** [comment]
Confirmed. Input nodes are active, my mistake.

**Comment by darioizzo** [comment]
Will close this ticket as I could not find the issue reported. I did find a problem with scaling up group mutations though, which is addressed in 5fd0d585e72c53d8d84ec25fc5d4788ce0741c3f.
